### ISSUE TITLE
Request samples/sec as training metric

### ISSUE THREAD

**Issue by Lokiiiiii** [tensorflow/models#10631]
# Prerequisites

Please answer the following question for yourself before submitting an issue.

- [x] I checked to make sure that this feature has not been requested already.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/orbit/controller.py

## 2. Describe the feature you request

This is a request for a batch size agnostic metric to measure training speed. I would like to propose reporting samples/sec in addition to the existing steps/sec. 

## 3. Additional context

This would aide in figuring out the largest batch size that will yield the highest performance. The current method of finding steps/sec and multiplying it with batch size is tedious. 

## 4. Are you willing to contribute it? (Yes or No)
Yes

**Comment by saberkun** [comment]
steps/sec * batch_size = samples/sec for a single task trainer. Is it good enough?
For multi-task trainer, this is non-trivial as different task can have different batch size.
Another problem is the definition of "sample". When we enable bucketization, the batch size per batch is changing. When we enable packing of examples in a sequence, the number of examples in one sample is changing. Given the complexity for different cases, it is better to implement the samples/sec as a metric returned by trainer. The controller just have best effort default metrics. You can add more through implementing different trainers.

**Comment by Lokiiiiii** [comment]
Thanks for that input @saberkun. Consider this design:

The [StandardTrainer](https://github.com/tensorflow/models/blob/master/orbit/standard_runner.py#L71) has [create_train_loop_fn()](https://github.com/tensorflow/models/blob/master/orbit/standard_runner.py#L109) which implements the training loop. Currently this is set to be [stateless](https://github.com/tensorflow/models/blob/master/orbit/standard_runner.py#L125). However, ['state-ful' implementations](https://github.com/tensorflow/models/blob/master/orbit/utils/loop_fns.py#L76-L79) of the training loop does exist in loop_fns.py. 

The [state](https://github.com/tensorflow/models/blob/master/orbit/utils/loop_fns.py#L60-L63) is a dictionary managed by a [callable](https://github.com/tensorflow/models/blob/master/orbit/utils/loop_fns.py#L60-L63) that is executed with current state and current outputs as arguments. I would like to implement a default state manager function that counts the number of outputs to obtain the current batch size. 

The [train_loop_begin()](https://github.com/tensorflow/models/blob/master/orbit/standard_runner.py#L149-L157) and [train_loop_end()](https://github.com/tensorflow/models/blob/master/orbit/standard_runner.py#L181-L193) can be modified to measure the train_loop_latency and calculate samples/second. The state will be passed from [train_loop_end() to the controller](https://github.com/tensorflow/models/blob/23ac769fd7beb0914123b230dfcb5c4da016cc47/orbit/controller.py#L439) where it can be [manipulated into logs and summaries](https://github.com/tensorflow/models/blob/23ac769fd7beb0914123b230dfcb5c4da016cc47/orbit/controller.py#L455-L463). 

All custom Trainers derived from StandardTrainer that have override the train_loop_begin/end functions will need to additionally call super().train_loop_begin/end to get the samples/second metric.

Here's what I could use some help with:

1. Why is the current StandardTrainer using the stateless design over the 'state-ful' implementation ?
2. There are differences in the implementation of the [stateless](https://github.com/tensorflow/models/blob/master/orbit/utils/loop_fns.py#L88) and ['state-ful'](https://github.com/tensorflow/models/blob/master/orbit/utils/loop_fns.py#L125) design when it comes to while loop training. Can these differences be reconciled ?

**Comment by saberkun** [comment]
Hi, @rxsang for the discussion. I don't think we have stateless assumption. TF is stateful in general.
See the subclass of StandardTrainer: https://github.com/tensorflow/models/blob/master/official/core/base_trainer.py#L195
You can add states and return a dictionary of tensors or arrays: https://github.com/tensorflow/models/blob/master/official/core/base_trainer.py#L327 for the train_loop_end. Essentially, the samples/sec fitting your setup can be returned here.
